### DISCUSSION TITLE
form works fine in development but not production

### DISCUSSION THREAD

**Discussion by behzadmehrabi** [react-hook-form/react-hook-form#5610] [Q&A]
Hi, I have a very large form and I'm handling it with `react-window`,
I have something like this: 
```js
  const methods = useForm({
    defaultValues: initialValues ?? {
      number: 0,
      date: '',
      type_id: 0,
      branch_id: 0,
      description: '',
      items: [createVoucherItem()],
    },
  })

  const { fields } = useFieldArray({
    control: methods.control,
    name: 'items',
  })
```

and one of my components is this:
```js
import React from 'react'
import { Controller, UseFormReturn } from 'react-hook-form'
import { InputCell } from './libVoucherForm'

type Props = {
  index: number
} & UseFormReturn<any>

const VoucherFormFollowupNumberCell = React.memo<Props>(
  ({ index, control}) => {
    return (
      <Controller
        name={`items.${index}.followupNumber`}
        control={control}
        render={({ field }) => (
          <InputCell tw="select-all" {...field} />
        )}
      />
    )
  },
  (prevProps, nextProps) => prevProps.formState.isDirty === nextProps.formState.isDirty
)
```

in development, it works fine, but in production the value the input getting is `undefined`, what's the problem?

**Comment by bluebill1049** [comment]
need more detail on this, which version and can you share a reproducible repo.

**Reply by behzadmehrabi** [reply]
Hi, I couldn't reproduce the problem that form works in dev but not production.
but I created a small repo for you where the `defaultValue` of a nested field doesn't work as expected. [github repo](https://github.com/Bezmehrabi/testing-rhf),
the `description` field works, but the `debtor` doesn't

version: 7.8.5

**Reply by bluebill1049** [reply]
when you say it doesn't work, you mean the submitted value is `undefined`?

**Reply by behzadmehrabi** [reply]
The submitted value is ok, but the input value is `undefiend` rather than being the defaultValue I've set.
I rolled back to version `7.7.1` and there is no such problem there.

**Reply by bluebill1049** [reply]
> but the input value is undefiend rather than the defaultValue I've set.

what do you mean value is `undefined`? as the input itself has the text "undefined"? what's the steps to reproduce the issue? I am clone the repo now.

**Reply by bluebill1049** [reply]
i see `undefined` even in dev build

**Reply by behzadmehrabi** [reply]
so, as you can see in `App.tsx`,
```js
const methods = useForm({
    defaultValues: {
      number: 0,
      date: '',
      type_id: 0,
      branch_id: 0,
      description: 'hello',
      items: [{...createVoucherItem(), debtor: '12345'}],
    },
  });
```
the `description` value is `hello` and `items.0.debtor` is `12345`.

when you run the application,
the `hello` value is in the `description` input as expected.
but the second input which is for `items.0.debtor` has no value in it. it should be `12345`.

**Reply by behzadmehrabi** [reply]
yeah I told, I don't know why it was working in dev for me in the real project, but here it's `undefined` in both `dev` and `prod`. but it shouldn't be, right?

**Reply by bluebill1049** [reply]
I am looking into it.

**Reply by bluebill1049** [reply]
is this a field array? you will need to provide `defaultValue`.

**Reply by bluebill1049** [reply]
> defaultValue must be set for all inputs. Supplied defaultValues in the useForm hook will prepare the fields object with default value.

https://react-hook-form.com/api/usefieldarray

**Reply by behzadmehrabi** [reply]
ooo yes, that was the problem, thank you.


one thing I want to add is,
with version `7.7.1`, my page is so much smoother and performant comparing to version `7.8.7`,
I'm not changing any of my code, just switching between these 2 versions makes the difference, why is that?
if you need any detail on that, please let me know.

**Reply by bluebill1049** [reply]
Oh really?? Any chance U can share a performance monitor and see which functions is lead to this performance downgrade. So I can investigate and fix.

**Reply by behzadmehrabi** [reply]
sure, I've monitored both versions with chrome.
[chrome-profiling.zip](https://github.com/react-hook-form/react-hook-form/files/6660659/chrome-profiling.zip)

will also upload react profiler result in the next few hours.

**Reply by bluebill1049** [reply]
thanks, also some more context, is the downgrade during user interaction or initial mount?

**Reply by behzadmehrabi** [reply]
Hi again,
sorry mate, I said few hours but it's been few days now xD,
the initial mount is so slow compare to `7.7.1`.

**Reply by behzadmehrabi** [reply]
it's a bit annoying to profile it because it hangs, but I'm trying.

**Reply by bluebill1049** [reply]
Thanks, I haven't got a chance to look into it. by the way, we are trying to deprecate the defaultValue at useFieldArray to avoid this problem that occurred.

https://github.com/react-hook-form/react-hook-form/releases/tag/v7.8.9-beta.1
https://github.com/react-hook-form/react-hook-form/releases/tag/v7.8.9-beta.0

**Reply by behzadmehrabi** [reply]
yeah, I saw that, great work :thumbsup:
